setup.sh

```diff
@@ -59,109 +83,157 @@ function clean_legacy()
 {
-    local kw_fish_path="set -gx PATH $PATH:$HOME/.config/kw"
+  # TODO: allow completion for local installation
+  if [[ "$system_wide" == true ]]; then
 
-    say "Fish detected. Setting up fish support."
-    echo_n_run mkdir -p "$FISH_COMPLETION_PATH"
-    echo_n_run rsync -vr $SRCDIR/kw.fish "$FISH_COMPLETION_PATH"/kw.fish
+    echo_n_run rsync -v "completion/bash_autocomplete.sh" "$bashcompletionpath"
 
-    if ! grep -F "$kw_fish_path" "$FISH_CONFIG_PATH"/config.fish; then
-       echo "$kw_fish_path" >> "$FISH_CONFIG_PATH"/config.fish
+    if command -v fish &> /dev/null; then
+      say "Fish detected. Setting up fish support."
+      mkdir -p "$(dirname "$fishcompletionpath")"
+      echo_n_run rsync -r "completion/kw.fish" "$fishcompletionpath"
     fi
+  fi
 }
 
-# Synchronize .vim and .vimrc with repository.
 function synchronize_files()
 {
   say "Installing ..."
 
-  echo_n_run mkdir -p "$INSTALLTO"
+  mkdir -p "$libdir" "$sharedir" "$datadir" "$cachedir"
+  mkdir -p "$(dirname "$binpath")"
 
-  # Copy the script
-  echo_n_run cp "$APPLICATIONNAME" "$INSTALLTO"
-  echo_n_run rsync -vr "$SRCDIR" "$INSTALLTO"
-  echo_n_run rsync -vr "$SOUNDS" "$INSTALLTO"
-  echo_n_run rsync -vr "$DOCUMENTATION" "$INSTALLTO"
+  echo_n_run cp "kw" "$binpath"
+  sed -i -e "s,##KW_INSTALL_PREFIX_TOKEN##,$prefix/,g" "$binpath"
+  echo_n_run rsync -r "src/" "$libdir"
 
-  # Configuration
-  echo_n_run rsync -vr "$CONFIG_DIR" "$INSTALLTO"
-  setup_config_file
+  echo_n_run rsync -r "sounds" "$sharedir"
+  echo_n_run rsync -r "etc/" "$sharedir"
+  echo_n_run rsync -r "documentation" "$sharedir"
 
-  if [ -f "$HOME/.bashrc" ]; then
-      # Add to bashrc
-      echo "# $APPLICATIONNAME" >> "$HOME/.bashrc"
-      echo "PATH=\$PATH:$INSTALLTO" >> "$HOME/.bashrc"
-      echo "source $INSTALLTO/$SRCDIR/$BASH_AUTOCOMPLETE.sh" >> "$HOME/.bashrc"
-  else
-      warning "Unable to find a shell."
+  if [[ "$system_wide" == true ]]; then
+    mkdir -p "$(dirname "$manpath")"
+    echo_n_run rst2man "documentation/man/kw.rst" "$manpath"
   fi
 
-  if command -v fish &> /dev/null; then
-      synchronize_fish
-  fi
-
-  # Create ~/kw for support some of the operations
-  mkdir -p "$KW_DIR"
+  setup_config_file
+  synchronize_completion_files
 }
 
-function install_home()
+function reinstall()
 {
-  # First clean old installation
   clean_legacy
-  # Synchronize of vimfiles
   synchronize_files
 }
 
+function wrong_usage()
+{
+  complain "Invalid number of arguments"
+  usage
+  exit 1
+}
+
+function set_prefix()
+{
+  prefix="$@"
+  if [[ "$prefix" == "/usr" ]] || [[ "$prefix" == "/usr/" ]]; then
+    system_wide="true"
+  else
+    system_wide="false"
+  fi
+}
+
+# This function receives all arguments of the script and check if the second
+# argument is "--prefix=<VALUE>". If so, set $prefix to <VALUE>. If the number
+# of arguments exceeds 2 or the second arg is not as expected, exit with an
+# error.
+function maybe_parse_prefix()
```

Comment:
Maybe I misunderstood what you want to achieve here, but I think you merge `set_prefix` and this function like this:

```
  for arg do
    shift
    if [[ "$arg" =~ ^(--prefix|-p) ]]; then
      # Do something
    fi
    set -- "$@" "$arg"
  done
```